gems/pending/VolumeManager/MiqLvm.rb

```diff
@@ -295,6 +302,436 @@ def parseString
+
+   'Q',                       'device_details_root',
+
+   'L',                       'data_block_size',     # in 512-byte sectors
+
+   'L',                       'metadata_block_size', # in 512-byte sectors
+   'Q',                       'metadata_nr_blocks',
+
+   'L',                       'compat_flags',
+   'L',                       'compat_ro_flags',
+   'L',                       'incompat_flags'
+  ])
+
+  SPACE_MAP = BinaryStruct.new([
+    'Q',                      'nr_blocks',
+    'Q',                      'nr_allocated',
+    'Q',                      'bitmap_root',
+    'Q',                      'ref_count_root'
+  ])
+
+  DISK_NODE = BinaryStruct.new([
+    'L',                      'csum',
+    'L',                      'flags',
+    'Q',                      'blocknr',
+
+    'L',                      'nr_entries',
+    'L',                      'max_entries',
+    'L',                      'value_size',
+    'L',                      'padding'
+    #'Q',                      'keys'
+  ])
+
+  INDEX_ENTRY = BinaryStruct.new([
+    'Q',                      'blocknr',
+    'L',                      'nr_free',
+    'L',                      'none_free_before'
+  ])
+
+  METADATA_INDEX = BinaryStruct.new([
+    'L',                      'csum',
+    'L',                      'padding',
+    'Q',                      'blocknr'
+  ])
+
+  BITMAP_HEADER = BinaryStruct.new([
+    'L',                      'csum',
+    'L',                      'notused',
+    'Q',                      'blocknr'
+  ])
+
+  DEVICE_DETAILS = BinaryStruct.new([
+    'Q',                      'mapped_blocks',
+    'Q',                      'transaction_id',
+    'L',                      'creation_time',
+    'L',                      'snapshotted_time'
+  ])
+
+  MAPPING_DETAILS = BinaryStruct.new([
+    'Q',                       'value'
+  ])
+
+  class BTree
+    FLAGS = { :internal => 1, :leaf => 2}
+
+    attr_accessor :root_address
+
+    def initialize(superblock, root_address, value_type)
+      @superblock   = superblock
+      @root_address = root_address
+      @value_type   = value_type
+    end
+
+    def root
+      @root ||= begin
+        @superblock.seek root_address
+        @superblock.read_struct DISK_NODE
+      end
+    end
+
+    def internal?
+      (root['flags'] & FLAGS[:internal]) != 0
+    end
+
+    def leaf?
+      (root['flags'] & FLAGS[:leaf]) != 0
+    end
+
+    def num_entries
+      @num_entries ||= root['nr_entries']
+    end
+
+    def max_entries
+      @max_entries ||= root['max_entries']
+    end
+
+    def key_address(i)
+      root_address + DISK_NODE.size + i * 8
+    end
+
+    def value_base
+      key_address(max_entries)
+    end
+
+    def value_address(i)
+      value_base + @value_type.size * i
+    end
+
+    def keys
+      @keys ||=
+        0.upto(num_entries - 1).collect do |i|
+          @superblock.seek key_address(i)
+          @superblock.read_64.unpack('Q').first
```

Comment:
From reading this it seems that all of the blocks are sequential...as such, I think instead of seek/read in a loop, it might be better to seek to the first position, read num_entries worth of blocks and unpack them all in one shot...if that's possible.